sasl_test.go

```diff
@@ -0,0 +1,205 @@
+		t.Fatal(err)
+	}
+
+	gotResp, err := saslXOAUTH2InitialResponse("someuser@example.com", "ya29.vF9dft4qmTc2Nvb3RlckBhdHRhdmlzdGEuY29tCg")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(wantedResp, gotResp) {
+		t.Errorf("Initial response does not match expected:\n %s", testDiff(gotResp, wantedResp))
+	}
+}
+
+//RFC6749 defines the OAUTH2 as comprising VSCHAR elements (\x20-7E)
+func TestSaslXOAUTH2InvalidBearer(t *testing.T) {
+	tests := []struct {
+		label   string
+		illegal string
+	}{
+		{
+			label:   "char outside range",
+			illegal: "illegalChar\x00",
+		},
+		{
+			label:   "empty bearer",
+			illegal: "",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.label, func(t *testing.T) {
+			_, err := saslXOAUTH2InitialResponse("someuser@example.com", tt.illegal)
+			if err == nil {
+				t.Errorf("Expected invalid bearer to be rejected")
+			}
+		})
+	}
+}
+
+// https://developers.google.com/gmail/imap/xoauth2-protocol gives no guidance on the formation of the username field
+// or even if the username is actually required. However, disallowing \x01 seems reasonable as this would interfere
+// with the encoding.
+func TestSaslXOAUTH2InvalidUsername(t *testing.T) {
+	_, err := saslXOAUTH2InitialResponse("illegalChar\x01Within", "ya29.vF9dft4qmTc2Nvb3RlckBhdHRhdmlzdGEuY29tCg")
+	if err == nil {
+		t.Errorf("Expected invalid username to be rejected")
+	}
+}
+
+func TestSaslXOAUTH2EmptyUsername(t *testing.T) {
+	_, err := saslXOAUTH2InitialResponse("", "ya29.vF9dft4qmTc2Nvb3RlckBhdHRhdmlzdGEuY29tCg")
+	if err != nil {
+		t.Errorf("Expected empty username to be accepted")
+	}
+}
+
+func TestConnSASLXOAUTH2AuthSuccess(t *testing.T) {
+	buf, err := peerResponse(
+		[]byte("AMQP\x03\x01\x00\x00"),
+		frame{
+			type_:   frameTypeSASL,
+			channel: 0,
+			body:    &saslMechanisms{Mechanisms: []symbol{saslMechanismXOAUTH2}},
+		},
+		frame{
+			type_:   frameTypeSASL,
+			channel: 0,
+			body:    &saslOutcome{Code: codeSASLOK},
+		},
+		[]byte("AMQP\x00\x01\x00\x00"),
+		frame{
+			type_:   frameTypeAMQP,
+			channel: 0,
+			body:    &performOpen{},
+		},
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c := testconn.New(buf)
+	client, err := New(c,
+		ConnSASLXOAUTH2("someuser@example.com", "ya29.vF9dft4qmTc2Nvb3RlckBhdHRhdmlzdGEuY29tCg", 512),
+		ConnIdleTimeout(10*time.Minute))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+}
+
+func TestConnSASLXOAUTH2AuthFail(t *testing.T) {
+	buf, err := peerResponse(
+		[]byte("AMQP\x03\x01\x00\x00"),
+		frame{
+			type_:   frameTypeSASL,
+			channel: 0,
+			body:    &saslMechanisms{Mechanisms: []symbol{saslMechanismXOAUTH2}},
+		},
+		frame{
+			type_:   frameTypeSASL,
+			channel: 0,
+			body:    &saslOutcome{Code: codeSASLAuth},
+		},
+	)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c := testconn.New(buf)
+	client, err := New(c,
+		ConnSASLXOAUTH2("someuser@example.com", "ya29.vF9dft4qmTc2Nvb3RlckBhdHRhdmlzdGEuY29tCg", 512),
+		ConnIdleTimeout(10*time.Minute))
+	if client != nil {
```

Comment:
Go tends to assume that if `err != nil` the other return values are invalid regardless of whether they're nil or not (unless documented otherwise). Can you change this to `if err != nil`?

Reply:
Done